// @vsintellicode/typescript-intellicode-plugin/index.js
/*! Copyright (c) Microsoft Corporation. All rights reserved. */
module.exports=function(e){var n={};function t(o){if(n[o])return n[o].exports;var i=n[o]={i:o,l:!1,exports:{}};return e[o].call(i.exports,i,i.exports,t),i.l=!0,i.exports}return t.m=e,t.c=n,t.d=function(e,n,o){t.o(e,n)||Object.defineProperty(e,n,{enumerable:!0,get:o})},t.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.t=function(e,n){if(1&n&&(e=t(e)),8&n)return e;if(4&n&&"object"==typeof e&&e&&e.__esModule)return e;var o=Object.create(null);if(t.r(o),Object.defineProperty(o,"default",{enumerable:!0,value:e}),2&n&&"string"!=typeof e)for(var i in e)t.d(o,i,function(n){return e[n]}.bind(null,i));return o},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return Object.prototype.hasOwnProperty.call(e,n)},t.p="",t(t.s=0)}([function(e,n,t){"use strict";var o=this&&this.__importStar||function(e){if(e&&e.__esModule)return e;var n={};if(null!=e)for(var t in e)Object.hasOwnProperty.call(e,t)&&(n[t]=e[t]);return n.default=e,n};const i=o(t(1)),r=o(t(3)),s=o(t(4)),a=o(t(5)),c=o(t(6)),u=1e6,l=1e9;let f,d,p=!1,m=500,y=void 0;function g(e,n,t){const o=S(...process.hrtime(e));n&&"None"===n.failureReason?t.metadata={type:n.foundNode.typeName,inIf:n.foundNode.isInConditional,patterns:n.sequences,computationTimeMS:o,failureReason:"None"}:t.metadata={computationTimeMS:o,failureReason:n?n.failureReason:"LookupKeyGenerationTimeout"}}function h(e,n){e.languageServiceHost.log&&e.languageServiceHost.log(`IntelliCode plugin: ${n}`)}function S(e,n){return(e*l+n)/u}const N=3.4,x=3.6,T=3.7;class k{constructor(e,n,t){this.ts=e,this.hostCancellationToken=n,this.timeoutMS=t,this.startTime=Date.now()}throwIfCancellationRequested(){if(this.isHostTokenCanceled())throw new v(!0);if(p&&this.isLocalTokenCanceled())throw new v(!1)}isHostTokenCanceled(){return this.hostCancellationToken.isCancellationRequested()}isLocalTokenCanceled(){return Math.abs(Date.now()-this.startTime)>=this.timeoutMS}}class v{constructor(e){this.canceledByHostToken=e}}e.exports=function(e){const n=e.typescript;let t;return{create:function(e){if(!n.versionMajorMinor)return e.languageService;const u=Number(n.versionMajorMinor);if(NaN===u||u<3.2)return e.languageService;const l=Object.create(null);for(let n of Object.keys(e.languageService)){const t=e.languageService[n];t&&(l[n]=t.bind(e.languageService))}function C(){if(!y&&f)try{const n=s.readFileSync(f),t=Buffer.from([214,105,172,23,156,35,36,45,137,37,195,176,189,110,127,78,69,171,205,160,3,202,114,220,252,109,233,221,252,181,45,47]),o=Buffer.from([60,12,128,217,163,220,176,146,213,18,190,59,212,88,192,21]),i=r.createDecipheriv("aes-256-cbc",t,o),c=Buffer.concat([i.update(n),i.final()]),u=a.gunzipSync(c);y=JSON.parse(u.toString("utf-8")),d="typescript_"+(y.Version||"unknown"),h(e,`successfully loaded model from ${f}`)}catch(n){h(e,`failed to load model from ${f}, exception: ${n}`)}return y}return o(e.config),t=e.languageServiceHost,h(e,`starting configuration: ${JSON.stringify(e.config)}`),l.getCompletionsAtPosition=(t,o,r)=>(function(e,n,t,o,r,s,a){const u=t.languageService.getCompletionsAtPosition(r,s,a);if(!u||!u.isMemberCompletion)return u;const l=t.project.getCancellationToken(),f=new k(e,l,m),y=process.hrtime(),C=t.languageService.getProgram(),b=C?C.getSourceFile(r):void 0;if(u.entries.length>0||!b||!c.isInString(e,b,s)){if(p){try{f.throwIfCancellationRequested();const e=O(f);g(y,e,u)}catch(n){if(!(n instanceof v))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException;g(y,void 0,u)}return u}const i=function(e){let n=void 0,i=[];const r=o(),s=r&&r.model;if(!s)return{recommendedWords:[],failureReason:"NoModelAvailable"};const a=O(e);if(!a.foundNode||!a.sequences)return{recommendedWords:[],failureReason:a.failureReason};const c=s[a.foundNode.typeName];if(c){h(t,"found some suggestion stuff");for(const o of a.sequences){e.throwIfCancellationRequested();const r=c[o],s=r&&(a.foundNode.isInConditional?r[1]:r[0]);if(s){n=o,h(t,`found suggestions for sequence ${o} : ${s.join(",")}`),i.push(...s.slice(0,5));break}}}return{typeName:a.foundNode.typeName,inConditional:a.foundNode.isInConditional,sequenceString:n,recommendedWords:i,failureReason:"None"}}(f),r=S(...process.hrtime(y));let s="None"===i.failureReason?"NotInModel":i.failureReason,a=[];i.recommendedWords.length>0&&(s="NotInIntersection",u.entries=u.entries.filter(e=>!e.replacementSpan||!e.insertText||"["!==e.insertText[0]),u.entries.some(e=>(function(e,n){const t=n.sortText;if(e<=N)return!t.endsWith("1");if(e<=x)return!t.endsWith("4");if(e<=T)return!t.endsWith("6");return!n.isFromUncheckedFile})(n,e))?i.recommendedWords.forEach((e,n)=>{const o=u.entries.findIndex(n=>n.name===e);if(o>=0){const e=u.entries[o];h(t,`marking recommendation ${e.name}`),s="None",I(e,n),a.push(e),u.entries.splice(o,1)}}):(h(t,"no strong recommendations, adding entries"),u.entries=u.entries.filter(e=>!i.recommendedWords.some(n=>e.name===n)),a=i.recommendedWords.map((n,t)=>{const o={name:n,kind:e.ScriptElementKind.memberVariableElement,sortText:"0"};return I(o,t),o})),u.entries=a.concat(u.entries));const c=function(e,n,t){return{Id:Date.now(),ModelVersion:d||"typescript_unknown",Count:0,FailureReason:e,Class:"",InIf:n.inConditional,ElapsedTime:t,Methods:[],PrevInvoc:"",ItemMetadata:Object.create(null)}}(s,i,r);i.typeName&&i.sequenceString&&(c.Count=a.length,c.ModelType=i.sequenceString&&"N"!==i.sequenceString?"Sequence":"Frequency",c.Class="NotInModel"!==s?i.typeName:"",c.Methods=a.map(e=>e.insertText),c.PrevInvoc=i.sequenceString,u.entries.forEach((e,n)=>{const o={Id:c.Id,Index:n,Method:n<a.length?e.insertText:""};c.ItemMetadata[e.name]=o,h(t,`adding item metadata ${e.name} => ${JSON.stringify(o)} to completion result`)})),h(t,`added metadata ${JSON.stringify(c)} to completion result`),u.metadata=c}function I(e,n){e.insertText=e.insertText||e.name,e.sortText=e.sortText||e.name,e.name="★ "+e.name,e.sortText=`*${String.fromCharCode("a".charCodeAt(0)+n)} ${e.sortText}`,h(t,`sort text for entry ${e.name} is ${e.sortText}`)}function O(n){const o=t.languageService.getProgram();if(!o)return{failureReason:"NoProgramFound"};h(t,"program found"),n.throwIfCancellationRequested();const a=o.getTypeChecker(),u=o.getSourceFile(r);if(!u)return{failureReason:"NoSourceFileFound"};h(t,"source file found"),n.throwIfCancellationRequested();const l=function(n,t,o){let r=c.findPrecedingToken(e,o,t);r&&o<=r.end&&(e.isIdentifier(r)||c.isKeyword(e,r.kind))&&(r=c.findPrecedingToken(e,r.getFullStart(),t));if(!r||r.kind!==e.SyntaxKind.DotToken)return;const s=r.parent;if(!e.isPropertyAccessExpression(s))return;const a=i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,n,s.expression,t).typeName;if(!a)return;return{isInConditional:A(s),node:s,typeName:a,sf:t}}(a,u,s);if(n.throwIfCancellationRequested(),!l)return{failureReason:"NoNodeFound"};h(t,`determined type: ${l.typeName}`);const f=function(o){const r=[];try{!function o(r,s,c,u){if(s.getStart()>c.getStart()||u.length>=2)return;n.throwIfCancellationRequested();const f=s.getChildren(l.sf);for(let e=f.length-1;e>=0;e--)o(r,f[e],c,u);if(e.isPropertyAccessExpression(s)&&s!=c){i.getTypeNameForPythia(e,i.GetTypeNameOptions.SkipAny,a,s.expression,l.sf).typeName===r&&s.name.text&&(h(t,`pushing prior access: ${s.name.text}`),u.push(s.name.text))}}(o,l.sf,l.node,r)}catch(n){if(!(n instanceof v))throw n;if(n.canceledByHostToken)throw new e.OperationCanceledException}h(t,`prior accesses: ${r.join(",")}`);const s=[];r.length>1&&s.push(r[1]+"~"+r[0]);r.length>0&&s.push("N~"+r[0]);return s.push("N"),h(t,`sequences: ${s.join(",")}`),s}(l.typeName);return h(t,`sequences: ${f.join(",")}`),{foundNode:l,sequences:f,failureReason:"None"}}function A(n){let t=n;for(;t.kind!==e.SyntaxKind.SourceFile;){const n=t.parent;if(e.isIfStatement(n)&&n.expression===t||e.isConditionalExpression(n)&&n.condition===t||e.isDoStatement(n)&&n.expression===t||e.isWhileStatement(n)&&n.expression===t||e.isForStatement(n)&&n.condition===t)return!0;t=n}return!1}return u})(n,u,e,C,t,o,r),l.getCompletionEntryDetails=(...t)=>(function(e,n,t,o,i,r,...s){const a=r.startsWith("★ "),c=a?r.slice("★ ".length):r;a&&h(n,`stripped name for recommendation: ${c}`);const u=(0,n.languageService.getCompletionEntryDetails)(o,i,c,...s);u&&a&&(u.name="★ "+u.name,h(n,`starred name for recommendation: ${u.name}`));return u})(n,e,C,...t),l},onConfigurationChanged:function(e){t&&t.log(`received configuration command: ${JSON.stringify(e)}`),o(e)}};function o(e){e.modelPath&&(f=e.modelPath),e.returnLookupKey&&(p=e.returnLookupKey),e.timeout&&(m=e.timeout)}}},function(e,n,t){"use strict";n.__esModule=!0;var o,i,r=t(2);!function(e){e[e.DontSkipAny=0]="DontSkipAny",e[e.SkipAny=1]="SkipAny"}(o=n.GetTypeNameOptions||(n.GetTypeNameOptions={})),function(e){e[e.SkipAny=0]="SkipAny",e[e.Any=1]="Any",e[e.AnyCallExpression=2]="AnyCallExpression",e[e.AnyPropertyAccessExpression=3]="AnyPropertyAccessExpression",e[e.AnyElementAccessExpression=4]="AnyElementAccessExpression",e[e.UnionOrIntersection=5]="UnionOrIntersection",e[e.OtherPrimitive=6]="OtherPrimitive",e[e.OtherNonPrimitive=7]="OtherNonPrimitive",e[e.Literal=8]="Literal",e[e.EmptyIdentifier=9]="EmptyIdentifier"}(i=n.SkipTypeReason||(n.SkipTypeReason={})),n.getTypeNameForPythia=function e(n,t,s,a,c){var u=s.getTypeAtLocation(a),l=u.aliasSymbol||u.getSymbol();if(l){if(0!=(l.flags&(n.SymbolFlags.TypeLiteral|n.SymbolFlags.ObjectLiteral)))return{typeName:s.typeToString(u)};var f=s.getFullyQualifiedName(l);if(f&&'"'===f[0]){var d=function(e){var n=e.indexOf('"',1);if(n>=1)return{partInQuotes:e.substring(1,n),rest:e.substring(n+1)};return{}}(f),p=d.partInQuotes,m=d.rest;if(p){var y=p;if(r.isAbsolute(y)){"index"===r.basename(y)&&(y=r.dirname(y));var g=y.lastIndexOf("/node_modules/");if(g>=0){var h=g+"/node_modules/".length;y=y.substring(h)}else y=r.basename(y);f='"'+y+'"'+(m||"")}}}return{typeName:f}}if(0==(u.getFlags()&n.TypeFlags.NonPrimitive)){if(0!=(u.flags&n.TypeFlags.Any)){var S;if(n.isCallExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"()"}:{skipReason:i.AnyCallExpression};if(n.isPropertyAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"."+a.name.text}:{skipReason:i.AnyPropertyAccessExpression};if(n.isElementAccessExpression(a))return(S=e(n,t,s,a.expression,c)).typeName?{typeName:S.typeName+"[]"}:{skipReason:i.AnyElementAccessExpression};if(t===o.SkipAny)return{skipReason:i.SkipAny};if(n.isIdentifier(a)){var N=a.text;return N?{typeName:N}:{skipReason:i.EmptyIdentifier}}return a.kind===n.SyntaxKind.ThisKeyword?{typeName:"this"}:{skipReason:i.Any}}return 0!=(u.flags&n.TypeFlags.StringLike)?{typeName:"string"}:0!=(u.flags&n.TypeFlags.Boolean)?{typeName:"boolean"}:0!=(u.flags&n.TypeFlags.NumberLike)?{typeName:"number"}:0!=(u.flags&n.TypeFlags.UnionOrIntersection)?{skipReason:i.UnionOrIntersection}:{skipReason:i.OtherPrimitive}}return{skipReason:i.OtherNonPrimitive}}},function(e,n){e.exports=require("path")},function(e,n){e.exports=require("crypto")},function(e,n){e.exports=require("fs")},function(e,n){e.exports=require("zlib")},function(e,n,t){"use strict";function o(e,n,t){return n.kind===e.SyntaxKind.EndOfFileToken?!!n.jsDoc:0!==n.getWidth(t)}function i(e,n){return e.isJsxText(n)&&(n.containsOnlyTriviaWhiteSpaces||n.containsOnlyWhiteSpaces)}function r(e,n,t,r){for(let s=t-1;s>=0;s--){if(i(e,n[s]));else if(o(e,n[s],r))return n[s]}}function s(e,n,t){if(a(e,n))return n;const o=n.getChildren(t),i=r(e,o,o.length,t);return i&&s(e,i,t)}function a(e,n){return e.isToken(n)&&!i(e,n)}Object.defineProperty(n,"__esModule",{value:!0}),n.isInString=function(e,n,t){const o=function(e,n,t){let o=n;e:for(;;){for(const i of o.getChildren(n)){if(i.getFullStart()>t)break;const n=i.getEnd();if(t<n||t===n&&i.kind===e.SyntaxKind.EndOfFileToken){o=i;continue e}}return o}}(e,n,t);if(o&&(e.isStringLiteral(o)||e.isTemplateLiteralToken(o))){const e=o.getStart(n),i=o.getEnd();if(e<t&&t<i)return!0;if(t===i)return!!o.isUnterminated}},n.findPrecedingToken=function(e,n,t,c,u){return function c(l){if(a(e,l)&&l.kind!==e.SyntaxKind.EndOfFileToken)return l;const f=l.getChildren(t);for(let a=0;a<f.length;a++){const l=f[a];if(n<l.end){if(l.getStart(t,!u)>=n||!o(e,l,t)||i(e,l)){const n=r(e,f,a,t);return n&&s(e,n,t)}return c(l)}}const d=r(e,f,f.length,t);return d&&s(e,d,t)}(c||t)},n.isKeyword=function(e,n){return e.SyntaxKind.FirstKeyword<=n&&n<=e.SyntaxKind.LastKeyword}}]);
// SIG // Begin signature block
// SIG // MIIjhQYJKoZIhvcNAQcCoIIjdjCCI3ICAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // +cHhA2ynpKhWoa0kUGWg1TEgQWsQhSzUOKWVhitSJsGg
// SIG // gg2BMIIF/zCCA+egAwIBAgITMwAAAd9r8C6Sp0q00AAA
// SIG // AAAB3zANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIwMTIxNTIxMzE0NVoX
// SIG // DTIxMTIwMjIxMzE0NVowdDELMAkGA1UEBhMCVVMxEzAR
// SIG // BgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1v
// SIG // bmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlv
// SIG // bjEeMBwGA1UEAxMVTWljcm9zb2Z0IENvcnBvcmF0aW9u
// SIG // MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
// SIG // trsZWRAAo6nx5LhcqAsHy9uaHyPQ2VireMBI9yQUOPBj
// SIG // 7dVLA7/N+AnKFFDzJ7P+grT6GkOE4cv5GzjoP8yQJ6yX
// SIG // ojEKkXti7HW/zUiNoF11/ZWndf8j1Azl6OBjcD416tSW
// SIG // Yvh2VfdW1K+mY83j49YPm3qbKnfxwtV0nI9H092gMS0c
// SIG // pCUsxMRAZlPXksrjsFLqvgq4rnULVhjHSVOudL/yps3z
// SIG // OOmOpaPzAp56b898xC+zzHVHcKo/52IRht1FSC8V+7QH
// SIG // TG8+yzfuljiKU9QONa8GqDlZ7/vFGveB8IY2ZrtUu98n
// SIG // le0WWTcaIRHoCYvWGLLF2u1GVFJAggPipwIDAQABo4IB
// SIG // fjCCAXowHwYDVR0lBBgwFgYKKwYBBAGCN0wIAQYIKwYB
// SIG // BQUHAwMwHQYDVR0OBBYEFDj2zC/CHZDRrQnzJlT7byOl
// SIG // WfPjMFAGA1UdEQRJMEekRTBDMSkwJwYDVQQLEyBNaWNy
// SIG // b3NvZnQgT3BlcmF0aW9ucyBQdWVydG8gUmljbzEWMBQG
// SIG // A1UEBRMNMjMwMDEyKzQ2MzAwOTAfBgNVHSMEGDAWgBRI
// SIG // bmTlUAXTgqoXNzcitW2oynUClTBUBgNVHR8ETTBLMEmg
// SIG // R6BFhkNodHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtp
// SIG // b3BzL2NybC9NaWNDb2RTaWdQQ0EyMDExXzIwMTEtMDct
// SIG // MDguY3JsMGEGCCsGAQUFBwEBBFUwUzBRBggrBgEFBQcw
// SIG // AoZFaHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3BraW9w
// SIG // cy9jZXJ0cy9NaWNDb2RTaWdQQ0EyMDExXzIwMTEtMDct
// SIG // MDguY3J0MAwGA1UdEwEB/wQCMAAwDQYJKoZIhvcNAQEL
// SIG // BQADggIBAJ56h7Q8mFBWlQJLwCtHqqup4aC/eUmULt0Z
// SIG // 6We7XUPPUEd/vuwPuIa6+1eMcZpAeQTm0tGCvjACxNNm
// SIG // rY8FoD3aWEOvFnSxq6CWR5G2XYBERvu7RExZd2iheCqa
// SIG // EmhjrJGV6Uz5wmjKNj16ADFTBqbEBELMIpmatyEN50UH
// SIG // wZSdD6DDHDf/j5LPGUy9QaD2LCaaJLenKpefaugsqWWC
// SIG // MIMifPdh6bbcmxyoNWbUC1JUl3HETJboD4BHDWSWoDxI
// SIG // D2J4uG9dbJ40QIH9HckNMyPWi16k8VlFOaQiBYj09G9s
// SIG // LMc0agrchqqZBjPD/RmszvHmqJlSLQmAXCUgcgcf6UtH
// SIG // EmMAQRwGcSTg1KsUl6Ehg75k36lCV57Z1pC+KJKJNRYg
// SIG // g2eI6clzkLp2+noCF75IEO429rjtujsNJvEcJXg74TjK
// SIG // 5x7LqYjj26Myq6EmuqWhbVUofPWm1EqKEfEHWXInppqB
// SIG // YXFpBMBYOLKc72DT+JyLNfd9utVsk2kTGaHHhrp+xgk9
// SIG // kZeud7lI/hfoPeHOtwIc0quJIXS+B5RSD9nj79vbJn1J
// SIG // x7RqusmBQy509Kv2Pg4t48JaBfBFpJB0bUrl5RVG05sK
// SIG // /5Qw4G6WYioS0uwgUw499iNC+Yud9vrh3M8PNqGQ5mJm
// SIG // JiFEjG2ToEuuYe/e64+SSejpHhFCaAFcMIIHejCCBWKg
// SIG // AwIBAgIKYQ6Q0gAAAAAAAzANBgkqhkiG9w0BAQsFADCB
// SIG // iDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0
// SIG // b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1p
// SIG // Y3Jvc29mdCBDb3Jwb3JhdGlvbjEyMDAGA1UEAxMpTWlj
// SIG // cm9zb2Z0IFJvb3QgQ2VydGlmaWNhdGUgQXV0aG9yaXR5
// SIG // IDIwMTEwHhcNMTEwNzA4MjA1OTA5WhcNMjYwNzA4MjEw
// SIG // OTA5WjB+MQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2Fz
// SIG // aGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UE
// SIG // ChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSgwJgYDVQQD
// SIG // Ex9NaWNyb3NvZnQgQ29kZSBTaWduaW5nIFBDQSAyMDEx
// SIG // MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA
// SIG // q/D6chAcLq3YbqqCEE00uvK2WCGfQhsqa+laUKq4Bjga
// SIG // BEm6f8MMHt03a8YS2AvwOMKZBrDIOdUBFDFC04kNeWSH
// SIG // fpRgJGyvnkmc6Whe0t+bU7IKLMOv2akrrnoJr9eWWcpg
// SIG // GgXpZnboMlImEi/nqwhQz7NEt13YxC4Ddato88tt8zpc
// SIG // oRb0RrrgOGSsbmQ1eKagYw8t00CT+OPeBw3VXHmlSSnn
// SIG // Db6gE3e+lD3v++MrWhAfTVYoonpy4BI6t0le2O3tQ5GD
// SIG // 2Xuye4Yb2T6xjF3oiU+EGvKhL1nkkDstrjNYxbc+/jLT
// SIG // swM9sbKvkjh+0p2ALPVOVpEhNSXDOW5kf1O6nA+tGSOE
// SIG // y/S6A4aN91/w0FK/jJSHvMAhdCVfGCi2zCcoOCWYOUo2
// SIG // z3yxkq4cI6epZuxhH2rhKEmdX4jiJV3TIUs+UsS1Vz8k
// SIG // A/DRelsv1SPjcF0PUUZ3s/gA4bysAoJf28AVs70b1FVL
// SIG // 5zmhD+kjSbwYuER8ReTBw3J64HLnJN+/RpnF78IcV9uD
// SIG // jexNSTCnq47f7Fufr/zdsGbiwZeBe+3W7UvnSSmnEyim
// SIG // p31ngOaKYnhfsi+E11ecXL93KCjx7W3DKI8sj0A3T8Hh
// SIG // hUSJxAlMxdSlQy90lfdu+HggWCwTXWCVmj5PM4TasIgX
// SIG // 3p5O9JawvEagbJjS4NaIjAsCAwEAAaOCAe0wggHpMBAG
// SIG // CSsGAQQBgjcVAQQDAgEAMB0GA1UdDgQWBBRIbmTlUAXT
// SIG // gqoXNzcitW2oynUClTAZBgkrBgEEAYI3FAIEDB4KAFMA
// SIG // dQBiAEMAQTALBgNVHQ8EBAMCAYYwDwYDVR0TAQH/BAUw
// SIG // AwEB/zAfBgNVHSMEGDAWgBRyLToCMZBDuRQFTuHqp8cx
// SIG // 0SOJNDBaBgNVHR8EUzBRME+gTaBLhklodHRwOi8vY3Js
// SIG // Lm1pY3Jvc29mdC5jb20vcGtpL2NybC9wcm9kdWN0cy9N
// SIG // aWNSb29DZXJBdXQyMDExXzIwMTFfMDNfMjIuY3JsMF4G
// SIG // CCsGAQUFBwEBBFIwUDBOBggrBgEFBQcwAoZCaHR0cDov
// SIG // L3d3dy5taWNyb3NvZnQuY29tL3BraS9jZXJ0cy9NaWNS
// SIG // b29DZXJBdXQyMDExXzIwMTFfMDNfMjIuY3J0MIGfBgNV
// SIG // HSAEgZcwgZQwgZEGCSsGAQQBgjcuAzCBgzA/BggrBgEF
// SIG // BQcCARYzaHR0cDovL3d3dy5taWNyb3NvZnQuY29tL3Br
// SIG // aW9wcy9kb2NzL3ByaW1hcnljcHMuaHRtMEAGCCsGAQUF
// SIG // BwICMDQeMiAdAEwAZQBnAGEAbABfAHAAbwBsAGkAYwB5
// SIG // AF8AcwB0AGEAdABlAG0AZQBuAHQALiAdMA0GCSqGSIb3
// SIG // DQEBCwUAA4ICAQBn8oalmOBUeRou09h0ZyKbC5YR4WOS
// SIG // mUKWfdJ5DJDBZV8uLD74w3LRbYP+vj/oCso7v0epo/Np
// SIG // 22O/IjWll11lhJB9i0ZQVdgMknzSGksc8zxCi1LQsP1r
// SIG // 4z4HLimb5j0bpdS1HXeUOeLpZMlEPXh6I/MTfaaQdION
// SIG // 9MsmAkYqwooQu6SpBQyb7Wj6aC6VoCo/KmtYSWMfCWlu
// SIG // WpiW5IP0wI/zRive/DvQvTXvbiWu5a8n7dDd8w6vmSiX
// SIG // mE0OPQvyCInWH8MyGOLwxS3OW560STkKxgrCxq2u5bLZ
// SIG // 2xWIUUVYODJxJxp/sfQn+N4sOiBpmLJZiWhub6e3dMNA
// SIG // BQamASooPoI/E01mC8CzTfXhj38cbxV9Rad25UAqZaPD
// SIG // XVJihsMdYzaXht/a8/jyFqGaJ+HNpZfQ7l1jQeNbB5yH
// SIG // PgZ3BtEGsXUfFL5hYbXw3MYbBL7fQccOKO7eZS/sl/ah
// SIG // XJbYANahRr1Z85elCUtIEJmAH9AAKcWxm6U/RXceNcbS
// SIG // oqKfenoi+kiVH6v7RyOA9Z74v2u3S5fi63V4GuzqN5l5
// SIG // GEv/1rMjaHXmr/r8i+sLgOppO6/8MO0ETI7f33VtY5E9
// SIG // 0Z1WTk+/gFcioXgRMiF670EKsT/7qMykXcGhiJtXcVZO
// SIG // SEXAQsmbdlsKgEhr/Xmfwb1tbWrJUnMTDXpQzTGCFVww
// SIG // ghVYAgEBMIGVMH4xCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xKDAm
// SIG // BgNVBAMTH01pY3Jvc29mdCBDb2RlIFNpZ25pbmcgUENB
// SIG // IDIwMTECEzMAAAHfa/AukqdKtNAAAAAAAd8wDQYJYIZI
// SIG // AWUDBAIBBQCgga4wGQYJKoZIhvcNAQkDMQwGCisGAQQB
// SIG // gjcCAQQwHAYKKwYBBAGCNwIBCzEOMAwGCisGAQQBgjcC
// SIG // ARUwLwYJKoZIhvcNAQkEMSIEIHss+67rvTtEPz9Mev3Y
// SIG // j26mlu7hX26ZMTGOEU7ZpLmXMEIGCisGAQQBgjcCAQwx
// SIG // NDAyoBSAEgBNAGkAYwByAG8AcwBvAGYAdKEagBhodHRw
// SIG // Oi8vd3d3Lm1pY3Jvc29mdC5jb20wDQYJKoZIhvcNAQEB
// SIG // BQAEggEAY0Q9F5kXcygtsiomPEoBR7WSH5GjpRDQ5wIV
// SIG // z+4cjE8Cdhr4LhtKkVm6YiJLy/Jo+iwYPvOrPK31utox
// SIG // p2mpBSCmt/zlpBqRfFmWraF/95knk5fCkBxH7Z9p8tST
// SIG // 4KxKaxugIhX8Lotvw63BZD1/fHVsIQ43gEan6FhfZYlm
// SIG // UXb3MvoP7Q8voXP29RE8+aDaq7QvbcIq/8sAWf0atC9W
// SIG // zvCWMJ1WQ8yZJTN4YGYvjiOGWOiTB/1RswHqF6Q2GHFg
// SIG // 8Znu/e/WiytDYRcXDN5h6h+HVlFPDGNHABEeUdqY3VKr
// SIG // ohXE+tf2OQnT4iiUNj8WKzEsMSbEL+cyS8g6uL4Wl6GC
// SIG // EuYwghLiBgorBgEEAYI3AwMBMYIS0jCCEs4GCSqGSIb3
// SIG // DQEHAqCCEr8wghK7AgEDMQ8wDQYJYIZIAWUDBAIBBQAw
// SIG // ggFRBgsqhkiG9w0BCRABBKCCAUAEggE8MIIBOAIBAQYK
// SIG // KwYBBAGEWQoDATAxMA0GCWCGSAFlAwQCAQUABCBsddvI
// SIG // XR4BSpo8RwRZjWp3TCbeh/EiZceOYLtS8MMoEgIGYK6W
// SIG // xT/FGBMyMDIxMDUyNjIxNDU1MS43NDVaMASAAgH0oIHQ
// SIG // pIHNMIHKMQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2Fz
// SIG // aGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UE
// SIG // ChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSUwIwYDVQQL
// SIG // ExxNaWNyb3NvZnQgQW1lcmljYSBPcGVyYXRpb25zMSYw
// SIG // JAYDVQQLEx1UaGFsZXMgVFNTIEVTTjpENkJELUUzRTct
// SIG // MTY4NTElMCMGA1UEAxMcTWljcm9zb2Z0IFRpbWUtU3Rh
// SIG // bXAgU2VydmljZaCCDj0wggTxMIID2aADAgECAhMzAAAB
// SIG // UFii1KebCzDrAAAAAAFQMA0GCSqGSIb3DQEBCwUAMHwx
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jv
// SIG // c29mdCBUaW1lLVN0YW1wIFBDQSAyMDEwMB4XDTIwMTEx
// SIG // MjE4MjYwM1oXDTIyMDIxMTE4MjYwM1owgcoxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xJTAjBgNVBAsTHE1pY3Jvc29mdCBB
// SIG // bWVyaWNhIE9wZXJhdGlvbnMxJjAkBgNVBAsTHVRoYWxl
// SIG // cyBUU1MgRVNOOkQ2QkQtRTNFNy0xNjg1MSUwIwYDVQQD
// SIG // ExxNaWNyb3NvZnQgVGltZS1TdGFtcCBTZXJ2aWNlMIIB
// SIG // IjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA53p/
// SIG // lHsqWcMnG1aaDfMzUgKp2ZeNPkY0U8de4LD71WDcbz3u
// SIG // j867kvyeV9pjjegUbfl2iQEySb5AoB71Yiaq1O0N/U9K
// SIG // k39Nnzk085AUdc4QdooS910mhKURPy9sEmu74C/I4TxY
// SIG // tHWfJ56nI/em1S2kbz7OwDV3gxd8aWTYFEii9hHoAXJk
// SIG // VLHvvdrlpPzWLI/GNxAr9qj50gjREqnUPeyUuCt0eT8x
// SIG // 0ghtsWL0US6fm/dBVSNiQkK9SAI/dlDZHtsCh20LOVXc
// SIG // iRPv74IOrcmXWsVuwFaIyKuSnHzN/kGoTSqbHTh5t+yM
// SIG // GJg9S307G9LVYVy424e6OE+nAiolmwIDAQABo4IBGzCC
// SIG // ARcwHQYDVR0OBBYEFIJVGR8ZRdutZC9DwQBDO7frn0V+
// SIG // MB8GA1UdIwQYMBaAFNVjOlyKMZDzQ3t8RhvFM2hahW1V
// SIG // MFYGA1UdHwRPME0wS6BJoEeGRWh0dHA6Ly9jcmwubWlj
// SIG // cm9zb2Z0LmNvbS9wa2kvY3JsL3Byb2R1Y3RzL01pY1Rp
// SIG // bVN0YVBDQV8yMDEwLTA3LTAxLmNybDBaBggrBgEFBQcB
// SIG // AQROMEwwSgYIKwYBBQUHMAKGPmh0dHA6Ly93d3cubWlj
// SIG // cm9zb2Z0LmNvbS9wa2kvY2VydHMvTWljVGltU3RhUENB
// SIG // XzIwMTAtMDctMDEuY3J0MAwGA1UdEwEB/wQCMAAwEwYD
// SIG // VR0lBAwwCgYIKwYBBQUHAwgwDQYJKoZIhvcNAQELBQAD
// SIG // ggEBAFE/aHRyjv7j49PAHi+T9h4sTDqhyt6D5PSQHZpY
// SIG // H9l7fMRe2PQR4hUDVmDCnfxyNq8mDrt0gALN7cWWkaWp
// SIG // b2NEdelliEMwvRvaKkg+0DTdETMbIogjDOuf56vcUjyK
// SIG // 5mIemk1xYCDLoeBizZO4DEBcm1G4NmZS2bJ/a+Uc8V/8
// SIG // fGite4IBFipooo9ZYG/+gBPmwD1LhMsudLmulPWbNPm4
// SIG // vPx752bOj4jMBobV8/7Ea2WgGI2y72glzpPg20bxwaGB
// SIG // 0EJt+UKz1CH4jXMN9IIiDEcu8WB44ppfyMWpDBVCeepi
// SIG // xN0r520hsWisKaT5ZCLMOR4XpnUhh7OUpa0E214wggZx
// SIG // MIIEWaADAgECAgphCYEqAAAAAAACMA0GCSqGSIb3DQEB
// SIG // CwUAMIGIMQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2Fz
// SIG // aGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UE
// SIG // ChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMTIwMAYDVQQD
// SIG // EylNaWNyb3NvZnQgUm9vdCBDZXJ0aWZpY2F0ZSBBdXRo
// SIG // b3JpdHkgMjAxMDAeFw0xMDA3MDEyMTM2NTVaFw0yNTA3
// SIG // MDEyMTQ2NTVaMHwxCzAJBgNVBAYTAlVTMRMwEQYDVQQI
// SIG // EwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRtb25kMR4w
// SIG // HAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRpb24xJjAk
// SIG // BgNVBAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1wIFBDQSAy
// SIG // MDEwMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKC
// SIG // AQEAqR0NvHcRijog7PwTl/X6f2mUa3RUENWlCgCChfvt
// SIG // fGhLLF/Fw+Vhwna3PmYrW/AVUycEMR9BGxqVHc4JE458
// SIG // YTBZsTBED/FgiIRUQwzXTbg4CLNC3ZOs1nMwVyaCo0UN
// SIG // 0Or1R4HNvyRgMlhgRvJYR4YyhB50YWeRX4FUsc+TTJLB
// SIG // xKZd0WETbijGGvmGgLvfYfxGwScdJGcSchohiq9LZIlQ
// SIG // YrFd/XcfPfBXday9ikJNQFHRD5wGPmd/9WbAA5ZEfu/Q
// SIG // S/1u5ZrKsajyeioKMfDaTgaRtogINeh4HLDpmc085y9E
// SIG // uqf03GS9pAHBIAmTeM38vMDJRF1eFpwBBU8iTQIDAQAB
// SIG // o4IB5jCCAeIwEAYJKwYBBAGCNxUBBAMCAQAwHQYDVR0O
// SIG // BBYEFNVjOlyKMZDzQ3t8RhvFM2hahW1VMBkGCSsGAQQB
// SIG // gjcUAgQMHgoAUwB1AGIAQwBBMAsGA1UdDwQEAwIBhjAP
// SIG // BgNVHRMBAf8EBTADAQH/MB8GA1UdIwQYMBaAFNX2VsuP
// SIG // 6KJcYmjRPZSQW9fOmhjEMFYGA1UdHwRPME0wS6BJoEeG
// SIG // RWh0dHA6Ly9jcmwubWljcm9zb2Z0LmNvbS9wa2kvY3Js
// SIG // L3Byb2R1Y3RzL01pY1Jvb0NlckF1dF8yMDEwLTA2LTIz
// SIG // LmNybDBaBggrBgEFBQcBAQROMEwwSgYIKwYBBQUHMAKG
// SIG // Pmh0dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9wa2kvY2Vy
// SIG // dHMvTWljUm9vQ2VyQXV0XzIwMTAtMDYtMjMuY3J0MIGg
// SIG // BgNVHSABAf8EgZUwgZIwgY8GCSsGAQQBgjcuAzCBgTA9
// SIG // BggrBgEFBQcCARYxaHR0cDovL3d3dy5taWNyb3NvZnQu
// SIG // Y29tL1BLSS9kb2NzL0NQUy9kZWZhdWx0Lmh0bTBABggr
// SIG // BgEFBQcCAjA0HjIgHQBMAGUAZwBhAGwAXwBQAG8AbABp
// SIG // AGMAeQBfAFMAdABhAHQAZQBtAGUAbgB0AC4gHTANBgkq
// SIG // hkiG9w0BAQsFAAOCAgEAB+aIUQ3ixuCYP4FxAz2do6Eh
// SIG // b7Prpsz1Mb7PBeKp/vpXbRkws8LFZslq3/Xn8Hi9x6ie
// SIG // JeP5vO1rVFcIK1GCRBL7uVOMzPRgEop2zEBAQZvcXBf/
// SIG // XPleFzWYJFZLdO9CEMivv3/Gf/I3fVo/HPKZeUqRUgCv
// SIG // OA8X9S95gWXZqbVr5MfO9sp6AG9LMEQkIjzP7QOllo9Z
// SIG // Kby2/QThcJ8ySif9Va8v/rbljjO7Yl+a21dA6fHOmWaQ
// SIG // jP9qYn/dxUoLkSbiOewZSnFjnXshbcOco6I8+n99lmqQ
// SIG // eKZt0uGc+R38ONiU9MalCpaGpL2eGq4EQoO4tYCbIjgg
// SIG // tSXlZOz39L9+Y1klD3ouOVd2onGqBooPiRa6YacRy5rY
// SIG // DkeagMXQzafQ732D8OE7cQnfXXSYIghh2rBQHm+98eEA
// SIG // 3+cxB6STOvdlR3jo+KhIq/fecn5ha293qYHLpwmsObvs
// SIG // xsvYgrRyzR30uIUBHoD7G4kqVDmyW9rIDVWZeodzOwjm
// SIG // mC3qjeAzLhIp9cAvVCch98isTtoouLGp25ayp0Kiyc8Z
// SIG // QU3ghvkqmqMRZjDTu3QyS99je/WZii8bxyGvWbWu3EQ8
// SIG // l1Bx16HSxVXjad5XwdHeMMD9zOZN+w2/XU/pnR4ZOC+8
// SIG // z1gFLu8NoFA12u8JJxzVs341Hgi62jbb01+P3nSISRKh
// SIG // ggLPMIICOAIBATCB+KGB0KSBzTCByjELMAkGA1UEBhMC
// SIG // VVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT
// SIG // B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jw
// SIG // b3JhdGlvbjElMCMGA1UECxMcTWljcm9zb2Z0IEFtZXJp
// SIG // Y2EgT3BlcmF0aW9uczEmMCQGA1UECxMdVGhhbGVzIFRT
// SIG // UyBFU046RDZCRC1FM0U3LTE2ODUxJTAjBgNVBAMTHE1p
// SIG // Y3Jvc29mdCBUaW1lLVN0YW1wIFNlcnZpY2WiIwoBATAH
// SIG // BgUrDgMCGgMVACMNe59y8TVrRhGbWpZZpDGWHqASoIGD
// SIG // MIGApH4wfDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldh
// SIG // c2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNV
// SIG // BAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEmMCQGA1UE
// SIG // AxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENBIDIwMTAw
// SIG // DQYJKoZIhvcNAQEFBQACBQDkWRVDMCIYDzIwMjEwNTI3
// SIG // MDI0MzE1WhgPMjAyMTA1MjgwMjQzMTVaMHgwPgYKKwYB
// SIG // BAGEWQoEATEwMC4wCgIFAORZFUMCAQAwCwIBAAIDBx3l
// SIG // AgH/MAcCAQACAhFfMAoCBQDkWmbDAgEAMDYGCisGAQQB
// SIG // hFkKBAIxKDAmMAwGCisGAQQBhFkKAwKgCjAIAgEAAgMH
// SIG // oSChCjAIAgEAAgMBhqAwDQYJKoZIhvcNAQEFBQADgYEA
// SIG // Ce3toB4ydq7MMV2xmxmnjur32chZ9VSAdsRXMKsx499U
// SIG // bJGOpk+g0SpTHzbXajcHABaKBkRGSPP8dWIpMR5r4vI3
// SIG // hw8eYo7IenSfvQKSEUYbUoi7LsjnGSVZNh0mMBtorF63
// SIG // DDwyN0jJXsEhDAxlVE1r51g5wXZvpDVsIWUAn9YxggMN
// SIG // MIIDCQIBATCBkzB8MQswCQYDVQQGEwJVUzETMBEGA1UE
// SIG // CBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEe
// SIG // MBwGA1UEChMVTWljcm9zb2Z0IENvcnBvcmF0aW9uMSYw
// SIG // JAYDVQQDEx1NaWNyb3NvZnQgVGltZS1TdGFtcCBQQ0Eg
// SIG // MjAxMAITMwAAAVBYotSnmwsw6wAAAAABUDANBglghkgB
// SIG // ZQMEAgEFAKCCAUowGgYJKoZIhvcNAQkDMQ0GCyqGSIb3
// SIG // DQEJEAEEMC8GCSqGSIb3DQEJBDEiBCA5aMRLh0L99gpb
// SIG // TcQFpZnfLA/TYWr/Ec+mcwT1+cXIbjCB+gYLKoZIhvcN
// SIG // AQkQAi8xgeowgecwgeQwgb0EIGz0Pp8vmsJOGBcmYYiF
// SIG // 1dLohjVmDqinjt2Qonwv8qP2MIGYMIGApH4wfDELMAkG
// SIG // A1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAO
// SIG // BgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29m
// SIG // dCBDb3Jwb3JhdGlvbjEmMCQGA1UEAxMdTWljcm9zb2Z0
// SIG // IFRpbWUtU3RhbXAgUENBIDIwMTACEzMAAAFQWKLUp5sL
// SIG // MOsAAAAAAVAwIgQggNp6+I3qjl5xPBe6OLmOg+ORLgzc
// SIG // P1E0nF5fH+8+uWwwDQYJKoZIhvcNAQELBQAEggEAzMJl
// SIG // Fe7cEXnTxEGxBO+l359sBUaeqFArkwX/a35N+zHQDmD/
// SIG // Mg1PtdUzxRFpBg0NO5sE+kxBb3E0DljJvgZ/vbAWdQ69
// SIG // FFmBGzC4UqQfwXYRIwX4PZckFmXsZvtkz1AzBnuprK9e
// SIG // /i1DR51YPrBcDE8v0+6dijBMsidmYY74EHsw1q6ujVGa
// SIG // GA/254XHEKyDNexZVzlQxmrdamxMC3oTkh+9LJl5E7YG
// SIG // YdXmxF3r1CcXMW89BZO5zOyv+LP2UG9JMsTL7yVUnMsv
// SIG // 3ifInTvFyP5Q9Ad3NxZDRM/pr0vN3t1AFxCCCGI015m2
// SIG // U1mvz57RAUMYd98hqd/b5qzzB6Uw1Q==
// SIG // End signature block
